File: app/controllers/AttachmentController.java
/*
 * Copyright (c) 2018 The members of the EXAM Consortium (https://confluence.csc.fi/display/EXAM/Konsortio-organisaatio)
 *
 * Licensed under the EUPL, Version 1.1 or - as soon they will be approved by the European Commission - subsequent
 * versions of the EUPL (the "Licence");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * https://joinup.ec.europa.eu/software/page/eupl/licence-eupl
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the Licence is distributed
 * on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and limitations under the Licence.
 */

package controllers;

import be.objectify.deadbolt.java.actions.Group;
import be.objectify.deadbolt.java.actions.Pattern;
import be.objectify.deadbolt.java.actions.Restrict;
import controllers.base.BaseController;
import io.ebean.DB;
import io.ebean.ExpressionList;
import java.io.File;
import java.io.IOException;
import java.util.concurrent.CompletionStage;
import javax.inject.Inject;
import models.Attachment;
import models.Comment;
import models.Exam;
import models.LanguageInspection;
import models.Role;
import models.User;
import models.api.AttachmentContainer;
import models.questions.EssayAnswer;
import models.questions.Question;
import models.sections.ExamSectionQuestion;
import org.apache.pekko.stream.IOResult;
import org.apache.pekko.stream.javadsl.FileIO;
import org.apache.pekko.stream.javadsl.Source;
import org.apache.pekko.util.ByteString;
import play.libs.Files;
import play.mvc.Http;
import play.mvc.Http.MultipartFormData.FilePart;
import play.mvc.Result;
import sanitizers.Attrs;
import security.Authenticated;
import util.config.ConfigReader;
import util.file.FileHandler;

public class AttachmentController extends BaseController implements LocalAttachmentInterface {

    @Inject
    private ConfigReader configReader;

    @Inject
    private FileHandler fileHandler;

    @Authenticated
    @Restrict({ @Group("STUDENT") })
    @Override
    public CompletionStage<Result> addAttachmentToQuestionAnswer(Http.Request request) {
        MultipartForm mf = getForm(request);
        FilePart<Files.TemporaryFile> filePart = mf.getFilePart();
        long qid = Long.parseLong(mf.getForm().get("questionId")[0]);

        // first check if answer already exists
        ExamSectionQuestion question = DB
            .find(ExamSectionQuestion.class)
            .fetch("essayAnswer")
            .where()
            .idEq(qid)
            .eq("examSection.exam.creator", request.attrs().get(Attrs.AUTHENTICATED_USER))
            .findOne();
        if (question == null) {
            return wrapAsPromise(forbidden());
        }
        if (question.getEssayAnswer() == null) {
            EssayAnswer answer = new EssayAnswer();
            question.setEssayAnswer(answer);
            question.save();
        }

        String newFilePath;
        try {
            newFilePath =
                copyFile(
                    filePart.getRef(),
                    "question",
                    Long.toString(qid),
                    "answer",
                    question.getEssayAnswer().getId().toString()
                );
        } catch (IOException e) {
            return wrapAsPromise(internalServerError("i18n_error_creating_attachment"));
        }
        // Remove existing one if found
        EssayAnswer answer = question.getEssayAnswer();
        fileHandler.removePrevious(question.getEssayAnswer());

        Attachment attachment = fileHandler.createNew(filePart, newFilePath);
        answer.setAttachment(attachment);
        answer.save();
        return wrapAsPromise(ok(answer));
    }

    private CompletionStage<Result> replaceAndFinish(
        AttachmentContainer ac,
        FilePart<Files.TemporaryFile> fp,
        String path
    ) {
        // Remove existing one if found
        fileHandler.removePrevious(ac);

        Attachment attachment = fileHandler.createNew(fp, path);

        ac.setAttachment(attachment);
        ac.save();

        return wrapAsPromise(ok(attachment));
    }

    @Restrict({ @Group("TEACHER"), @Group("ADMIN") })
    @Override
    public CompletionStage<Result> addAttachmentToQuestion(Http.Request request) {
        MultipartForm mf = getForm(request);
        long qid = Long.parseLong(mf.getForm().get("questionId")[0]);
        Question question = DB
            .find(Question.class)
            .fetch("examSectionQuestions.examSection.exam.parent")
            .where()
            .idEq(qid)
            .findOne();
        if (question == null) {
            return wrapAsPromise(notFound());
        }
        FilePart<Files.TemporaryFile> filePart = mf.getFilePart();
        String newFilePath;
        try {
            newFilePath = copyFile(filePart.getRef(), "question", Long.toString(qid));
        } catch (IOException e) {
            return wrapAsPromise(internalServerError("i18n_error_creating_attachment"));
        }
        return replaceAndFinish(question, filePart, newFilePath);
    }

    @Restrict({ @Group("TEACHER"), @Group("ADMIN") })
    @Override
    public Result deleteQuestionAttachment(Long id) {
        Question question = DB.find(Question.class, id);
        if (question == null) {
            return notFound();
        }
        fileHandler.removePrevious(question);
        return ok();
    }

    @Authenticated
    @Restrict({ @Group("ADMIN"), @Group("STUDENT") })
    @Override
    public CompletionStage<Result> deleteQuestionAnswerAttachment(Long qid, Http.Request request) {
        User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
        ExamSectionQuestion question;
        if (user.hasRole(Role.Name.STUDENT)) {
            question =
                DB.find(ExamSectionQuestion.class).where().idEq(qid).eq("examSection.exam.creator", user).findOne();
        } else {
            question = DB.find(ExamSectionQuestion.class, qid);
        }
        if (
            question != null && question.getEssayAnswer() != null && question.getEssayAnswer().getAttachment() != null
        ) {
            EssayAnswer answer = question.getEssayAnswer();
            Attachment aa = answer.getAttachment();
            answer.setAttachment(null);
            answer.save();
            fileHandler.removeAttachmentFile(aa.getFilePath());
            aa.delete();
            return wrapAsPromise(ok(answer));
        }
        return wrapAsPromise(notFound());
    }

    @Authenticated
    @Restrict({ @Group("TEACHER"), @Group("ADMIN") })
    @Override
    public CompletionStage<Result> deleteExamAttachment(Long id, Http.Request request) {
        Exam exam = DB.find(Exam.class, id);
        User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
        if (exam == null) {
            return wrapAsPromise(notFound());
        }
        if (!user.hasRole(Role.Name.ADMIN) && !exam.isOwnedOrCreatedBy(user)) {
            return wrapAsPromise(forbidden("i18n_error_access_forbidden"));
        }

        fileHandler.removePrevious(exam);
        return wrapAsPromise(ok());
    }

    @Restrict({ @Group("TEACHER"), @Group("ADMIN") })
    @Override
    public CompletionStage<Result> deleteFeedbackAttachment(Long id, Http.Request request) {
        Exam exam = DB.find(Exam.class, id);
        if (exam == null) {
            return wrapAsPromise(notFound("i18n_exam_not_found"));
        }
        Comment comment = exam.getExamFeedback();
        fileHandler.removePrevious(comment);
        return wrapAsPromise(ok());
    }

    @Authenticated
    @Pattern(value = "CAN_INSPECT_LANGUAGE")
    @Override
    public CompletionStage<Result> deleteStatementAttachment(Long id, Http.Request request) {
        LanguageInspection inspection = DB.find(LanguageInspection.class).where().eq("exam.id", id).findOne();
        if (inspection == null || inspection.getStatement() == null) {
            return wrapAsPromise(notFound("i18n_exam_not_found"));
        }
        Comment comment = inspection.getStatement();
        fileHandler.removePrevious(comment);
        return wrapAsPromise(ok());
    }

    @Authenticated
    @Restrict({ @Group("TEACHER"), @Group("ADMIN") })
    @Override
    public CompletionStage<Result> addAttachmentToExam(Http.Request request) {
        MultipartForm mf = getForm(request);
        long eid = Long.parseLong(mf.getForm().get("examId")[0]);
        Exam exam = DB.find(Exam.class, eid);
        if (exam == null) {
            return wrapAsPromise(notFound());
        }
        User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
        if (!user.hasRole(Role.Name.ADMIN) && !exam.isOwnedOrCreatedBy(user)) {
            return wrapAsPromise(forbidden("i18n_error_access_forbidden"));
        }
        String newFilePath;
        FilePart<Files.TemporaryFile> filePart = mf.getFilePart();
        try {
            newFilePath = copyFile(filePart.getRef(), "exam", Long.toString(eid));
        } catch (IOException e) {
            return wrapAsPromise(internalServerError("i18n_error_creating_attachment"));
        }
        return replaceAndFinish(exam, filePart, newFilePath);
    }

    @Authenticated
    @Restrict({ @Group("TEACHER"), @Group("ADMIN") })
    @Override
    public CompletionStage<Result> addFeedbackAttachment(Long id, Http.Request request) {
        Exam exam = DB.find(Exam.class, id);
        if (exam == null) {
            return wrapAsPromise(notFound());
        }
        if (exam.getExamFeedback() == null) {
            Comment comment = new Comment();
            comment.setCreatorWithDate(request.attrs().get(Attrs.AUTHENTICATED_USER));
            comment.save();
            exam.setExamFeedback(comment);
            exam.update();
        }
        String newFilePath;
        FilePart<Files.TemporaryFile> filePart = getForm(request).getFilePart();
        try {
            newFilePath = copyFile(filePart.getRef(), "exam", id.toString(), "feedback");
        } catch (IOException e) {
            return wrapAsPromise(internalServerError("i18n_error_creating_attachment"));
        }
        Comment comment = exam.getExamFeedback();
        return replaceAndFinish(comment, filePart, newFilePath);
    }

    @Authenticated
    @Pattern(value = "CAN_INSPECT_LANGUAGE")
    @Override
    public CompletionStage<Result> addStatementAttachment(Long id, Http.Request request) {
        LanguageInspection inspection = DB.find(LanguageInspection.class).where().eq("exam.id", id).findOne();
        if (inspection == null) {
            return wrapAsPromise(notFound());
        }
        if (inspection.getStatement() == null) {
            Comment comment = new Comment();
            comment.setCreatorWithDate(request.attrs().get(Attrs.AUTHENTICATED_USER));
            comment.save();
            inspection.setStatement(comment);
            inspection.update();
        }
        FilePart<Files.TemporaryFile> filePart = getForm(request).getFilePart();
        String newFilePath;
        try {
            newFilePath = copyFile(filePart.getRef(), "exam", id.toString(), "inspectionstatement");
        } catch (IOException e) {
            return wrapAsPromise(internalServerError("i18n_error_creating_attachment"));
        }
        Comment comment = inspection.getStatement();
        return replaceAndFinish(comment, filePart, newFilePath);
    }

    @Authenticated
    @Restrict({ @Group("TEACHER"), @Group("ADMIN"), @Group("STUDENT") })
    @Override
    public CompletionStage<Result> downloadQuestionAttachment(Long id, Http.Request request) {
        User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
        Question question;
        if (user.hasRole(Role.Name.STUDENT)) {
            question =
                DB
                    .find(Question.class)
                    .where()
                    .idEq(id)
                    .eq("examSectionQuestions.examSection.exam.creator", user)
                    .findOne();
        } else {
            question = DB.find(Question.class, id);
        }
        if (question == null || question.getAttachment() == null) {
            return wrapAsPromise(notFound());
        }
        return serveAttachment(question.getAttachment());
    }

    @Authenticated
    @Restrict({ @Group("TEACHER"), @Group("ADMIN"), @Group("STUDENT") })
    @Override
    public CompletionStage<Result> downloadQuestionAnswerAttachment(Long qid, Http.Request request) {
        // TODO: Authorization?
        ExamSectionQuestion question = getExamSectionQuestion(request, qid);
        if (
            question == null || question.getEssayAnswer() == null || question.getEssayAnswer().getAttachment() == null
        ) {
            return wrapAsPromise(notFound());
        }
        return serveAttachment(question.getEssayAnswer().getAttachment());
    }

    @Restrict({ @Group("TEACHER"), @Group("ADMIN"), @Group("STUDENT") })
    @Override
    public CompletionStage<Result> downloadExamAttachment(Long id, Http.Request request) {
        // TODO: Authorization?
        Exam exam = DB.find(Exam.class, id);
        if (exam == null || exam.getAttachment() == null) {
            return wrapAsPromise(notFound());
        }
        return serveAttachment(exam.getAttachment());
    }

    @Authenticated
    @Restrict({ @Group("TEACHER"), @Group("ADMIN"), @Group("STUDENT") })
    @Override
    public CompletionStage<Result> downloadFeedbackAttachment(Long id, Http.Request request) {
        User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
        Exam exam;
        if (user.hasRole(Role.Name.STUDENT)) {
            exam = DB.find(Exam.class).where().idEq(id).eq("creator", user).findOne();
        } else {
            exam = DB.find(Exam.class, id);
        }
        if (exam == null || exam.getExamFeedback() == null || exam.getExamFeedback().getAttachment() == null) {
            return wrapAsPromise(notFound());
        }
        return serveAttachment(exam.getExamFeedback().getAttachment());
    }

    @Authenticated
    @Restrict({ @Group("TEACHER"), @Group("ADMIN"), @Group("STUDENT") })
    @Override
    public CompletionStage<Result> downloadStatementAttachment(Long id, Http.Request request) {
        User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
        ExpressionList<Exam> query = DB
            .find(Exam.class)
            .where()
            .idEq(id)
            .isNotNull("languageInspection.statement.attachment");
        if (user.hasRole(Role.Name.STUDENT)) {
            query = query.eq("creator", user);
        }
        Exam exam = query.findOne();
        if (exam == null) {
            return wrapAsPromise(notFound());
        }
        return serveAttachment(exam.getLanguageInspection().getStatement().getAttachment());
    }

    @Override
    public ConfigReader getConfigReader() {
        return configReader;
    }

    private CompletionStage<Result> serveAttachment(Attachment attachment) {
        File file = new File(attachment.getFilePath());
        if (!file.exists()) {
            return wrapAsPromise(internalServerError("i18n_file_not_found_but_referred_in_database"));
        }
        final Source<ByteString, CompletionStage<IOResult>> source = FileIO.fromPath(file.toPath());
        return serveAsBase64Stream(attachment, source);
    }

    private ExamSectionQuestion getExamSectionQuestion(Http.Request request, Long id) {
        User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
        if (user.hasRole(Role.Name.STUDENT)) {
            return DB.find(ExamSectionQuestion.class).where().idEq(id).eq("examSection.exam.creator", user).findOne();
        }
        return DB.find(ExamSectionQuestion.class, id);
    }

    private String copyFile(Files.TemporaryFile srcFile, String... pathParams) throws IOException {
        String newFilePath = fileHandler.createFilePath(pathParams);
        fileHandler.copyFile(srcFile, new File(newFilePath));
        return newFilePath;
    }
}

File: app/controllers/attachment/AttachmentController.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package controllers.attachment

import controllers.base.scala.ExamBaseController
import io.ebean.DB
import miscellaneous.config.ConfigReader
import miscellaneous.file.FileHandler
import miscellaneous.scala.DbApiHelper
import models.assessment.{Comment, LanguageInspection}
import models.attachment.{Attachment, AttachmentContainer}
import models.exam.Exam
import models.questions.{EssayAnswer, Question}
import models.sections.ExamSectionQuestion
import models.user.{Permission, Role}
import org.apache.pekko.stream.scaladsl.{FileIO, Source}
import org.apache.pekko.stream.{IOResult, Materializer}
import org.apache.pekko.util.ByteString
import play.api.libs.Files.TemporaryFile
import play.api.mvc.*
import play.api.mvc.MultipartFormData.FilePart
import security.scala.Auth.{AuthenticatedAction, authorized}
import security.scala.{Auth, AuthExecutionContext, PermissionFilter}

import java.io.File
import javax.inject.Inject
import scala.concurrent.Future
import scala.util.{Failure, Success, Try}

class AttachmentController @Inject()(
    val controllerComponents: ControllerComponents,
    val authenticated: AuthenticatedAction,
    val configReader: ConfigReader,
    val fileHandler: FileHandler,
    implicit val ec: AuthExecutionContext,
    implicit val mat: Materializer
) extends BaseController
    with ExamBaseController
    with DbApiHelper:

  def addAttachmentToQuestionAnswer(): Action[MultipartFormData[TemporaryFile]] =
    authenticated
      .andThen(authorized(Seq(Role.Name.STUDENT)))
      .async(parse.multipartFormData) { request =>
        parseMultipartForm(request) match
          case None => Future.successful(BadRequest("Invalid form data"))
          case Some((filePart, formData)) =>
            formData.get("questionId").flatMap(_.headOption).map(_.toLong) match
              case None => Future.successful(BadRequest("Missing questionId"))
              case Some(qid) =>
                val user = request.attrs(Auth.ATTR_USER)
                DB.find(classOf[ExamSectionQuestion])
                  .fetch("essayAnswer")
                  .where()
                  .idEq(qid)
                  .eq("examSection.exam.creator", user)
                  .find match
                  case None           => Future.successful(Forbidden)
                  case Some(question) =>
                    // Ensure essay answer exists
                    if Option(question.getEssayAnswer).isEmpty then
                      val answer = new EssayAnswer()
                      question.setEssayAnswer(answer)
                      question.save()

                    Try(
                      copyFile(filePart.ref, "question", qid.toString, "answer", question.getEssayAnswer.getId.toString)
                    ) match
                      case Failure(_) =>
                        Future.successful(InternalServerError("i18n_error_creating_attachment"))
                      case Success(newFilePath) =>
                        val answer = question.getEssayAnswer
                        fileHandler.removePrevious(answer)
                        val attachment = fileHandler.createNew(filePart.filename, filePart.contentType.getOrElse("application/octet-stream"), newFilePath)
                        answer.setAttachment(attachment)
                        answer.save()
                        Future.successful(Ok(answer.asJson))
      }

  def addAttachmentToQuestion(): Action[MultipartFormData[TemporaryFile]] =
    authenticated
      .andThen(authorized(Seq(Role.Name.TEACHER, Role.Name.ADMIN, Role.Name.SUPPORT)))
      .async(parse.multipartFormData) { request =>
        parseMultipartForm(request) match
          case None => Future.successful(BadRequest("Invalid form data"))
          case Some((filePart, formData)) =>
            formData.get("questionId").flatMap(_.headOption).map(_.toLong) match
              case None => Future.successful(BadRequest("Missing questionId"))
              case Some(qid) =>
                DB.find(classOf[Question])
                  .fetch("examSectionQuestions.examSection.exam.parent")
                  .where()
                  .idEq(qid)
                  .find match
                  case None => Future.successful(NotFound)
                  case Some(question) =>
                    Try(copyFile(filePart.ref, "question", qid.toString)) match
                      case Failure(_) =>
                        Future.successful(InternalServerError("i18n_error_creating_attachment"))
                      case Success(newFilePath) =>
                        replaceAndFinish(question, filePart, newFilePath)
      }

  def deleteQuestionAttachment(id: Long): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.TEACHER, Role.Name.ADMIN, Role.Name.SUPPORT))) { _ =>
      Option(DB.find(classOf[Question], id)) match
        case None => NotFound
        case Some(question) =>
          fileHandler.removePrevious(question)
          Ok
    }

  def deleteQuestionAnswerAttachment(qid: Long): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN, Role.Name.STUDENT, Role.Name.SUPPORT))).async { request =>
      val user = request.attrs(Auth.ATTR_USER)
      val question =
        if user.hasRole(Role.Name.STUDENT) then
          DB.find(classOf[ExamSectionQuestion])
            .where()
            .idEq(qid)
            .eq("examSection.exam.creator", user)
            .find
        else Option(DB.find(classOf[ExamSectionQuestion], qid))

      question match
        case Some(q) if Option(q.getEssayAnswer).exists(a => Option(a.getAttachment).isDefined) =>
          val answer = q.getEssayAnswer
          val aa     = answer.getAttachment
          answer.setAttachment(null)
          answer.save()
          fileHandler.removeAttachmentFile(aa.getFilePath)
          aa.delete()
          Future.successful(Ok(answer.asJson))
        case _ => Future.successful(NotFound)
    }

  def deleteExamAttachment(id: Long): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.TEACHER, Role.Name.ADMIN, Role.Name.SUPPORT))).async { request =>
      Option(DB.find(classOf[Exam], id)) match
        case None => Future.successful(NotFound)
        case Some(exam) =>
          val user = request.attrs(Auth.ATTR_USER)
          if !user.hasRole(Role.Name.ADMIN) && !exam.isOwnedOrCreatedBy(user) then
            Future.successful(Forbidden("i18n_error_access_forbidden"))
          else
            fileHandler.removePrevious(exam)
            Future.successful(Ok)
    }

  def deleteFeedbackAttachment(id: Long): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.TEACHER, Role.Name.ADMIN, Role.Name.SUPPORT))).async { _ =>
      Option(DB.find(classOf[Exam], id)) match
        case None => Future.successful(NotFound("i18n_exam_not_found"))
        case Some(exam) =>
          Option(exam.getExamFeedback).foreach(fileHandler.removePrevious)
          Future.successful(Ok)
    }

  def deleteStatementAttachment(id: Long): Action[AnyContent] =
    authenticated
      .andThen(PermissionFilter(Permission.Type.CAN_INSPECT_LANGUAGE))
      .async { _ =>
        DB.find(classOf[LanguageInspection]).where().eq("exam.id", id).find match
          case None => Future.successful(NotFound("i18n_exam_not_found"))
          case Some(inspection) =>
            Option(inspection.getStatement).foreach(fileHandler.removePrevious)
            Future.successful(Ok)
      }

  def addAttachmentToExam(): Action[MultipartFormData[TemporaryFile]] =
    authenticated
      .andThen(authorized(Seq(Role.Name.TEACHER, Role.Name.ADMIN, Role.Name.SUPPORT)))
      .async(parse.multipartFormData) { request =>
        parseMultipartForm(request) match
          case None => Future.successful(BadRequest("Invalid form data"))
          case Some((filePart, formData)) =>
            formData.get("examId").flatMap(_.headOption).map(_.toLong) match
              case None => Future.successful(BadRequest("Missing examId"))
              case Some(eid) =>
                Option(DB.find(classOf[Exam], eid)) match
                  case None => Future.successful(NotFound)
                  case Some(exam) =>
                    val user = request.attrs(Auth.ATTR_USER)
                    if !user.hasRole(Role.Name.ADMIN) && !exam.isOwnedOrCreatedBy(user) then
                      Future.successful(Forbidden("i18n_error_access_forbidden"))
                    else
                      Try(copyFile(filePart.ref, "exam", eid.toString)) match
                        case Failure(_) =>
                          Future.successful(InternalServerError("i18n_error_creating_attachment"))
                        case Success(newFilePath) =>
                          replaceAndFinish(exam, filePart, newFilePath)
      }

  def addFeedbackAttachment(id: Long): Action[MultipartFormData[TemporaryFile]] =
    authenticated
      .andThen(authorized(Seq(Role.Name.TEACHER, Role.Name.ADMIN, Role.Name.SUPPORT)))
      .async(parse.multipartFormData) { request =>
        Option(DB.find(classOf[Exam], id)) match
          case None => Future.successful(NotFound)
          case Some(exam) =>
            if Option(exam.getExamFeedback).isEmpty then
              val comment = new Comment()
              comment.setCreatorWithDate(request.attrs(Auth.ATTR_USER))
              comment.save()
              exam.setExamFeedback(comment)
              exam.update()

            parseMultipartForm(request) match
              case None => Future.successful(BadRequest("Invalid form data"))
              case Some((filePart, _)) =>
                Try(copyFile(filePart.ref, "exam", id.toString, "feedback")) match
                  case Failure(_) =>
                    Future.successful(InternalServerError("i18n_error_creating_attachment"))
                  case Success(newFilePath) =>
                    replaceAndFinish(exam.getExamFeedback, filePart, newFilePath)
      }

  def addStatementAttachment(id: Long): Action[MultipartFormData[TemporaryFile]] =
    authenticated
      .andThen(PermissionFilter(Permission.Type.CAN_INSPECT_LANGUAGE))
      .async(parse.multipartFormData) { request =>
        DB.find(classOf[LanguageInspection]).where().eq("exam.id", id).find match
          case None => Future.successful(NotFound)
          case Some(inspection) =>
            if Option(inspection.getStatement).isEmpty then
              val comment = new Comment()
              comment.setCreatorWithDate(request.attrs(Auth.ATTR_USER))
              comment.save()
              inspection.setStatement(comment)
              inspection.update()

            parseMultipartForm(request) match
              case None => Future.successful(BadRequest("Invalid form data"))
              case Some((filePart, _)) =>
                Try(copyFile(filePart.ref, "exam", id.toString, "inspectionstatement")) match
                  case Failure(_) =>
                    Future.successful(InternalServerError("i18n_error_creating_attachment"))
                  case Success(newFilePath) =>
                    replaceAndFinish(inspection.getStatement, filePart, newFilePath)
      }

  def downloadQuestionAttachment(id: Long): Action[AnyContent] =
    authenticated.async { request =>
      val user = request.attrs(Auth.ATTR_USER)
      val question =
        if user.hasRole(Role.Name.STUDENT) then
          DB.find(classOf[Question])
            .where()
            .idEq(id)
            .eq("examSectionQuestions.examSection.exam.creator", user)
            .find
        else Option(DB.find(classOf[Question], id))

      question match
        case Some(q) if Option(q.getAttachment).isDefined =>
          serveAttachment(q.getAttachment)
        case _ => Future.successful(NotFound)
    }

  def downloadQuestionAnswerAttachment(qid: Long): Action[AnyContent] =
    authenticated.async { request =>
      getExamSectionQuestion(request, qid) match
        case Some(q) if Option(q.getEssayAnswer).exists(a => Option(a.getAttachment).isDefined) =>
          serveAttachment(q.getEssayAnswer.getAttachment)
        case _ => Future.successful(NotFound)
    }

  def downloadExamAttachment(id: Long): Action[AnyContent] =
    authenticated.async { _ =>
      Option(DB.find(classOf[Exam], id)) match
        case Some(exam) if Option(exam.getAttachment).isDefined =>
          serveAttachment(exam.getAttachment)
        case _ => Future.successful(NotFound)
    }

  def downloadFeedbackAttachment(id: Long): Action[AnyContent] =
    authenticated.async { request =>
      val user = request.attrs(Auth.ATTR_USER)
      val exam =
        if user.hasRole(Role.Name.STUDENT) then DB.find(classOf[Exam]).where().idEq(id).eq("creator", user).find
        else Option(DB.find(classOf[Exam], id))

      exam match
        case Some(e) if Option(e.getExamFeedback).exists(f => Option(f.getAttachment).isDefined) =>
          serveAttachment(e.getExamFeedback.getAttachment)
        case _ => Future.successful(NotFound)
    }

  def downloadStatementAttachment(id: Long): Action[AnyContent] =
    authenticated.async { request =>
      val user  = request.attrs(Auth.ATTR_USER)
      val query = DB.find(classOf[Exam]).where().idEq(id).isNotNull("languageInspection.statement.attachment")
      val exam =
        if user.hasRole(Role.Name.STUDENT) then query.eq("creator", user).find
        else query.find

      exam match
        case Some(e) => serveAttachment(e.getLanguageInspection.getStatement.getAttachment)
        case None    => Future.successful(NotFound)
    }

  // Helper methods
  private def replaceAndFinish(
      ac: AttachmentContainer,
      fp: FilePart[TemporaryFile],
      path: String
  ): Future[Result] =
    Future {
      fileHandler.removePrevious(ac)
      val attachment = fileHandler.createNew(fp.filename, fp.contentType.getOrElse("application/octet-stream"), path)
      ac.setAttachment(attachment)
      ac.save()
      Ok(attachment.asJson)
    }

  private def serveAttachment(attachment: Attachment): Future[Result] =
    val file = new File(attachment.getFilePath)
    if !file.exists() then Future.successful(InternalServerError("i18n_file_not_found_but_referred_in_database"))
    else
      val source: Source[ByteString, Future[IOResult]] = FileIO.fromPath(file.toPath)
      serveAsBase64Stream(attachment, source)

  private def serveAsBase64Stream(
      attachment: Attachment,
      source: Source[ByteString, Future[IOResult]]
  ): Future[Result] =
    // Convert to base64 and stream
    source
      .map(_.encodeBase64)
      .runFold(ByteString.empty)(_ ++ _)
      .map { base64Data =>
        Ok(base64Data.utf8String)
          .withHeaders(
            "Content-Disposition" -> s"attachment; filename=\"${attachment.getFileName}\"",
            "Content-Type"        -> "application/octet-stream"
          )
      }

  private def getExamSectionQuestion(request: Request[?], id: Long): Option[ExamSectionQuestion] =
    val user = request.attrs(Auth.ATTR_USER)
    if user.hasRole(Role.Name.STUDENT) then
      DB.find(classOf[ExamSectionQuestion]).where().idEq(id).eq("examSection.exam.creator", user).find
    else Option(DB.find(classOf[ExamSectionQuestion], id))

  private def copyFile(srcFile: TemporaryFile, pathParams: String*): String =
    val newFilePath = fileHandler.createFilePath(pathParams*)
    fileHandler.copyFile(srcFile, new File(newFilePath))
    newFilePath

  private def parseMultipartForm(
      request: Request[MultipartFormData[TemporaryFile]]
  ): Option[(FilePart[TemporaryFile], Map[String, Seq[String]])] =
    request.body.file("file").map { filePart =>
      (filePart, request.body.dataParts)
    }
